runtime/compiler/control/JITaaSCompilationThread.cpp

```diff
@@ -822,43 +833,80 @@ bool handleServerMessage(JITServer::ClientStream *client, TR_J9VM *fe)
+            if (!thunkStart)
+               compInfoPT->getCompilation()->failCompilation<TR::CodeCacheError>("Failed to allocate space in the code cache");
+
+            thunkAddress = thunkStart + 8;
+            void *vmHelper = j9ThunkVMHelperFromSignature(fe->_jitConfig, signature.size(), &signature[0]);
+            compInfoPT->reloRuntime()->reloTarget()->performThunkRelocation(reinterpret_cast<uint8_t *>(thunkAddress), (UDATA)vmHelper);
             }
-         // Cannot recreate thunk on the client because code cache is not yet allocated.
-         // Save it until compilation ends and relocateThunks is called
-         compInfoPT->addThunkToBeRelocated(serializedThunk, signature);
-         client->write(response, JITServer::Void());
+         else
+            {
+            // For AOT, set address to received string, because it will be stored to SCC, so
+            // no need for code cache allocation
+            thunkAddress = reinterpret_cast<void *>(&serializedThunk[0] + 8);
+            }
+
+         // Ideally, should use signature.data() here, but setJ2IThunk has non-const pointer
+         // as argument, and it uses it to invoke a VM function that also takes non-const pointer.
+         thunkAddress = fe->setJ2IThunk(&signature[0], signature.size(), thunkAddress, comp);
+
+         client->write(response, thunkAddress);
+         }
+         break;
+      case MessageType::VM_needsInvokeExactJ2IThunk:
+         {
+         auto recv = client->getRecvData<std::string>();
+         std::string signature = std::get<0>(recv);
+
+         TR_J2IThunkTable *thunkTable = comp->getPersistentInfo()->getInvokeExactJ2IThunkTable();
+         // Ideally, should use signature.data() here, but findThunk takes non-const pointer
+         TR_J2IThunk *thunk = thunkTable->findThunk(&signature[0], fe);
+         client->write(response, thunk == NULL);
          }
          break;
       case MessageType::VM_setInvokeExactJ2IThunk:
          {
          auto recv = client->getRecvData<std::string>();
          std::string &serializedThunk = std::get<0>(recv);
 
-         if (comp->compileRelocatableCode())
+         uint8_t *thunkStart;
+         if (!comp->compileRelocatableCode())
             {
-            // For AOT, store thunk to SCC immediately, in case relocations get delayed
-            static_cast<TR_J9SharedCacheVM *>(fe)->persistJ2IThunk(reinterpret_cast<void *>(&serializedThunk[0]));
+            // Same as regular thunks, copy thunk to code cache for regular compilation,
+            // save to SCC for AOT compilation.
+            // Do not need relocation here, because helper address should have been originally
+            // fetched from the client.
+            thunkStart = TR_JITaaSRelocationRuntime::copyDataToCodeCache(serializedThunk.data(), serializedThunk.size(), fe);
+            if (!thunkStart)
```

Comment:
we could fail this compilation, but we have added an entry to our set of thunks. better to add after we receive conformation that the registration was successful at the client.